ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/behaviorTree/action/actions/SceneActionNodeExecutor.java

```diff
@@ -41,27 +45,72 @@ public void triggerExecution()
    @Override
    public void updateCurrentlyExecuting()
    {
+      state.setElapsedExecutionTime(timer.getElapsedTime());
+
       double timeout = 5.0;
       if (!timer.isRunning(timeout))
       {
+         state.getLogger().error("Timed out after %.1f s without finding a suitable detection.".formatted(timeout));
          state.setFailed(true);
          state.setIsExecuting(false);
          return;
       }
 
+      boolean printDebug = throttler.run();
+
       // Find a close stable detection
       PersistentDetection bestDetection = null;
       double closestDistanceSquared = Double.MAX_VALUE;
       IsaacROSFoundationPoseObject objectType = definition.getObjectType();
       cameraPosition.set(syncedRobot.getFramePoseReadOnly(HumanoidReferenceFrames::getExperimentalCameraFrame).getTranslation());
       for (PersistentDetection detection : scene.getPersistentDetections())
       {
-         if (!detection.getDetectedObjectClass().equals(objectType.yoloClass))
-            continue;
+         if (detection.getMostRecentDetection() instanceof IsaacROSFoundationPoseInstantDetection fpDetection)
+         {
+            if (!definition.getUseFoundationPose())
+            {
+               if (printDebug)
+                  state.getLogger().warn("Need YOLO but found FoundationPose name: %s".formatted(fpDetection.getObject().name()));
+
+               continue;
+            }
+
+            if (fpDetection.getObject() != objectType)
+            {
+               if (printDebug)
+                  state.getLogger().warn("Need object class type {} but found {}", objectType.name(), fpDetection.getObject().name());
+
+               continue;
+            }
+         }
+         else if (detection.getMostRecentDetection() instanceof YOLOv8InstantDetection yoloDetection)
+         {
+            if (definition.getUseFoundationPose())
+            {
+               if (printDebug)
+                  state.getLogger().warn("Need FoundationPose but found YOLOv8 name: %s"
+                                               .formatted(detection.getMostRecentDetection().getDetectedObjectClass()));
+
+               continue;
+            }
+
+            if (!yoloDetection.getDetectedObjectClass().equals(objectType.yoloClass))
+            {
+               if (printDebug)
+                  state.getLogger().warn("Need object class type {} but found {}", objectType.yoloClass, yoloDetection.getDetectedObjectClass());
+
+               continue;
+            }
+         }
```

Comment:
Above I fixed major bugs. I tested this on real robot and it works perfectly now.